### DISCUSSION TITLE
Convert string of a named expression in sympy to the expression itself

### DISCUSSION THREAD

**Discussion by sancho-s** [sympy/sympy#24483] [Q&A]
I have the following code

    import sympy as sp
    a, b = sp.symbols('a,b', real=True, positive=True)
    expr2 = 1.01 * a**1.01 * b**0.99
    print(type(expr2), '->', expr2)

Now I want a function that takes the string `'expr2'` and returns the expression `1.01 * a**1.01 * b**0.99`.
The ultimate objective is to put together the strings for two different expressions `'expr2'` and `'expr3'`, which should presumably give the same result, and verify their ratio, as in 

    def verify_ratio(vstr1, vstr2):
        """Compare the result of two different computations of the same quantity"""
        ratio = sp.N(sp.parsing.sympy_parser.parse_expr(vstr1)) / sp.parsing.sympy_parser.parse_expr(vstr2)
        print(vstr1 + ' / ' + vstr2, '=', sp.N(ratio))
        return

which does not work, as per what I tried:

    expr2 = 1.01 * a**1.01 * b**0.99
    print(type(expr2), '->', expr2)
        
    expr2b = sp.parsing.sympy_parser.parse_expr('expr2')
    print(type(expr2b), '->', expr2b)
    
    expr2c = sp.N(sp.parsing.sympy_parser.parse_expr('expr2'))
    print(type(expr2c), '->', expr2c)
    #print(sp.N(sp.parsing.sympy_parser.parse_expr('expr2')))
    
    expr2d = sp.sympify('expr2')
    print(type(expr2d), '->', expr2d)

with output

    <class 'sympy.core.mul.Mul'> -> 1.01*a**1.01*b**0.99
    <class 'sympy.core.symbol.Symbol'> -> expr2
    <class 'sympy.core.symbol.Symbol'> -> expr2
    <class 'sympy.core.symbol.Symbol'> -> expr2

None of my attempts achieved the objective.
Questions or links which did not help (at least for me):

 1. https://stackoverflow.com/questions/33606667/from-string-to-sympy-expression
 2. https://docs.sympy.org/latest/tutorials/intro-tutorial/basic_operations.html
 3. https://docs.sympy.org/latest/modules/parsing.html
 4. https://docs.sympy.org/latest/modules/core.html#sympy.core.sympify.sympify
 5. https://docs.sympy.org/latest/tutorials/intro-tutorial/manipulation.html

**Note**:
Besides the practical aspects of my objective, I don't know if there is any formal difference between `Symbol` (which is a specific class) and *expression*. From the sources I read (e.g., [this][1]) I did not arrive to a conclusion.
This understanding may help in solving the question.

  [1]: https://docs.sympy.org/latest/tutorials/intro-tutorial/manipulation.html

**Comment by oscarbenjamin** [comment]
What you are looking for is Python's `eval` function:
```python
In [1]: import sympy as sp
   ...: a, b = sp.symbols('a,b', real=True, positive=True)
   ...: expr2 = 1.01 * a**1.01 * b**0.99
   ...: print(type(expr2), '->', expr2)
<class 'sympy.core.mul.Mul'> -> 1.01*a**1.01*b**0.99

In [2]: eval('expr2')
Out[2]: 
      1.01  0.99
1.01⋅a    ⋅b  
```
https://docs.python.org/3/library/functions.html#eval

I definitely recommend taking a different approach though. I'm sure there's a better way to achieve your actual goal.

Taking a step back why do you want a function that takes a string and returns a global variable?

Better approaches most likely include using dome kind of data structure e.g. with a dict:
```python
expressions = {
    'expr2':  1.01 * a**1.01 * b**0.99,
    ...
}
```
Then you can get the expression with `expressions['expr2']`.

**Comment by sylee957** [comment]
I think that the confusion arises because of some misunderstanding between python operational semantics and sympy expressions.
For example, the statement `x = Symbol('x'); x = 2` doesn't mean that you can refer back the value of `x` from `Symbol('x')`, and would have any possible way to work because sympy is just a python library.

However, it is possible to achieve something like this indirectly by solving the equation trick.
For example, the imperative statement 
```
expr1 = 1.01 * a**1.01 * b**0.99
expr2 = expr1
expr3 = expr2
```

can also be defined as algebraic equation like below 
(I may leave the huge part of the work to write down the parser)

```python3
from sympy import *

a, b = symbols('a b', real=True, positive=True)
e1 = Symbol('expr1')
e2 = Symbol('expr2')
e3 = Symbol('expr3')

sol = solve(
    [Eq(e1, 1.01 * a**1.01 * b**0.99), Eq(e2, e1), Eq(e3, e2)], 
    [e1, e2, e3], dict=True)
```

And when you solve such equation, you can find which variables are assigned to which expressions by substitution
`e3.xreplace(sol[0])`.
However, this is not the entirely best way because algebraic equations way too much 'heavy' than using more dedicated solver for imperative calculus (like unifier), but if you need some tricks how to achieve that in sympy, I hope it may be helpful.
